pkg/coordinator/redemptions.go

```diff
@@ -0,0 +1,328 @@
+	"github.com/keep-network/keep-core/pkg/tbtc"
+	"math/big"
+	"sort"
+	"time"
+
+	"github.com/keep-network/keep-core/internal/hexutils"
+)
+
+type redemptionEntry struct {
+	walletPublicKeyHash [20]byte
+
+	redemptionKey        string
+	redeemerOutputScript bitcoin.Script
+	requestedAt          time.Time
+}
+
+// FindPendingRedemptions finds pending redemptions requests.
+func FindPendingRedemptions(
+	chain Chain,
+	walletPublicKeyHash [20]byte,
+	maxNumberOfDeposits uint16,
+) ([20]byte, []bitcoin.Script, error) {
+	logger.Infof("redemption max size: %d", maxNumberOfDeposits)
+
+	redemptionRequestMinAge, err := chain.GetRedemptionRequestMinAge()
+	if err != nil {
+		return [20]byte{}, nil, fmt.Errorf(
+			"failed to get redemption request minimum age: [%w]",
+			err,
+		)
+	}
+
+	_, _, _, _, redemptionTimeout, _, _, err := chain.GetRedemptionParameters()
+	if err != nil {
+		return [20]byte{}, nil, fmt.Errorf(
+			"failed to get redemption parameters: [%w]",
+			err,
+		)
+	}
+
+	getPendingRedemptionsFromWallet := func(walletToSweep [20]byte) ([]*redemptionEntry, error) {
+		pendingRedemptions, err := getPendingRedemptions(
+			chain,
+			walletToSweep,
+			int(maxNumberOfDeposits),
+			redemptionTimeout,
+			redemptionRequestMinAge,
+		)
+		if err != nil {
+			return nil,
+				fmt.Errorf(
+					"failed to get deposits for [%s] wallet: [%w]",
+					hexutils.Encode(walletToSweep[:]),
+					err,
+				)
+		}
+		return pendingRedemptions, nil
+	}
+
+	var redemptionsToPropose []*redemptionEntry
+	// If walletPublicKeyHash is not provided we need to find a wallet that has
+	// pending redemptions.
+	if walletPublicKeyHash == [20]byte{} {
+		walletRegisteredEvents, err := chain.PastNewWalletRegisteredEvents(nil)
+		if err != nil {
+			return [20]byte{}, nil, fmt.Errorf("failed to get registered wallets: [%w]", err)
+		}
+
+		// Take the oldest first
+		sort.SliceStable(walletRegisteredEvents, func(i, j int) bool {
+			return walletRegisteredEvents[i].BlockNumber < walletRegisteredEvents[j].BlockNumber
+		})
+
+		redeemingWallets := walletRegisteredEvents
+		// Only two the most recently created wallets are accepting redemptions.
+		if len(walletRegisteredEvents) >= 2 {
+			redeemingWallets = walletRegisteredEvents[len(walletRegisteredEvents)-2:]
+		}
+
+		for _, registeredWallet := range redeemingWallets {
+			logger.Infof(
+				"fetching pending redemption requests from wallet [%s]...",
+				hexutils.Encode(registeredWallet.WalletPublicKeyHash[:]),
+			)
+
+			pendingRedemptions, err := getPendingRedemptionsFromWallet(
+				registeredWallet.WalletPublicKeyHash,
+			)
+			if err != nil {
+				return [20]byte{}, nil, err
+			}
+
+			// Check if there are any unswept deposits in this wallet. If so
+			// sweep this wallet and don't check the other wallet.
+			if len(pendingRedemptions) > 0 {
+				walletPublicKeyHash = registeredWallet.WalletPublicKeyHash
+				redemptionsToPropose = pendingRedemptions
+				break
```

Comment:
Keep in mind when implementing the automation that this algorithm is optimal for sweeps when only the two most recently created wallets are sweeping but is not optimal for redemptions where all wallets can have pending redemption requests. If the maintainer is going to be triggered every 3 hours, it can't stop looking for redemption requests on the first found occurrence. It needs to propose more redemptions if needed, probably with some limit on the number of wallets that can sign at the same time.

Reply:
Yeah, that's clear.